src/AWS.Deploy.ServerMode.Client/ServerModeSession.cs

```diff
@@ -0,0 +1,277 @@
+                // If the error code other than -100, this is an unexpected exit code.
+                if (startServerTask.Result != TCP_PORT_ERROR)
+                {
+                    throw new InternalServerModeException($"\"{command}\" failed for unknown reason.");
+                }
+            }
+
+            throw new PortUnavailableException($"Free port unavailable in {_startPort}-{_endPort} range.");
+        }
+
+        public bool TryGetRestAPIClient(Func<Task<AWSCredentials>> credentialsGenerator, out IRestAPIClient? restApiClient)
+        {
+            if (_baseUrl == null || _aes == null)
+            {
+                restApiClient = null;
+                return false;
+            }
+
+            var httpClient = ServerModeHttpClientFactory.ConstructHttpClient(credentialsGenerator, _aes);
+            restApiClient = new RestAPIClient(_baseUrl, httpClient);
+            return true;
+        }
+
+        public bool TryGetDeploymentCommunicationClient(out IDeploymentCommunicationClient? deploymentCommunicationClient)
+        {
+            if (_baseUrl == null || _aes == null)
+            {
+                deploymentCommunicationClient = null;
+                return false;
+            }
+
+            deploymentCommunicationClient = new DeploymentCommunicationClient(_baseUrl);
+            return true;
+        }
+
+        public async ValueTask<bool> IsAlive(CancellationToken cancellationToken)
+        {
+            if (_baseUrl == null)
+            {
+                return false;
+            }
+
+            using var client = new HttpClient();
+
+            try
+            {
+                var response = await client.GetAsync(HealthUrl, cancellationToken);
+                return response.IsSuccessStatusCode;
+            }
+            catch (Exception)
+            {
+                return false;
+            }
+        }
+
+        #region Private methods
+
+        private Task<bool> IsServerAvailable(string baseUrl, CancellationToken cancellationToken)
+        {
+            return WaitUntilHelper.WaitUntilSuccessStatusCode(HealthUrl, TimeSpan.FromMilliseconds(100), TimeSpan.FromMinutes(10), cancellationToken);
+        }
+
+        private async Task<int> RunCommand(string command, params string[] stdIn)
+        {
+            var arguments = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? $"/c {command}" : $"-c \"{command}\"";
+
+            var processStartInfo = new ProcessStartInfo
+            {
+                FileName = GetSystemShell(),
+                Arguments = arguments,
+                UseShellExecute = false, // UseShellExecute must be false in allow redirection of StdIn.
+                RedirectStandardInput = true,
+                CreateNoWindow = !_diagnosticLoggingEnabled, // It allows displaying stdout and stderr on the screen.
+            };
+
+            var process = Process.Start(processStartInfo);
```

Comment:
The Toolkit will need `IServerModeSession` to have a "Stop" function, allowing it to terminate the server/process. The CLI makes an attempt to shut down by watching the parent process, but this should be a fallback mechanism. We are already seeing a negative impact from this during Toolkit development.

What would be the downside to storing `process` (or even just the processId), so that a `Stop` function could attempt to kill that process?

Reply:
I would keep this feature out of this PR. There are multiple ways to solve this problem
- Abort the process using process id
- Create an API endpoint which allows aborting the server

Let's first get this PR and then we can implement it later.

Reply:
Agreed. I've stubbed around this in the Toolkit for now in anticipation of something here in the near future.